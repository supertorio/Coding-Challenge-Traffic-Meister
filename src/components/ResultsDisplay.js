import React, {Fragment, PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {withStyles} from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import Chip from '@material-ui/core/Chip';
import Divider from '@material-ui/core/Divider';
import Grid from "@material-ui/core/Grid";
import Typography from '@material-ui/core/Typography';

import {DEFAULT_SELECT_VALUE, SELECTION_CATEGORIES} from "../constants/meister-consts";
import {
    getCurrentVehicleBrand,
    getCurrentVehicleColor,
    getCurrentVehicleType,
    getMeisterFormHasSelections,
    getCurrentVehicleImage
} from "../reducers/index-reducer";
import {clearSelection, resetVehicleForm} from '../actions/meister-form-actions';

const styles = theme => ({
    selections: {
        marginTop: 20
    },
    divider: {
        marginBottom: 20
    },
    chip: {
        textTransform: 'Capitalize',
        margin: theme.spacing.unit,
    },
    chipGroup: {
        paddingBottom: 10,
    },
    button: {
        margin: theme.spacing.unit,
    },
    vehicleImage: {
        maxWidth: 300,
        maxHeight: 200
    }
});

export class ResultsDisplay extends PureComponent {
    render() {
        const {classes, selectedType, selectedBrand, selectedColor, vehicleImage} = this.props;
        return (
            <div className={classes.selections}>
                {this.props.formHasSelections &&
                <Fragment>
                    <Divider className={classes.divider} />
                    <Grid container spacing={24}>
                        <Grid item xs={12} md={6}>
                            <Typography component="h2" variant="h5" gutterBottom>Your Selections</Typography>
                            <div className={classes.chipGroup}>
                                {selectedType !== DEFAULT_SELECT_VALUE &&
                                <Chip label={selectedType} className={classes.chip}
                                      onDelete={() => this.props.clearSelection(SELECTION_CATEGORIES.TYPE)}/>}
                                {selectedBrand !== DEFAULT_SELECT_VALUE &&
                                <Chip label={selectedBrand} className={classes.chip}
                                      onDelete={() => this.props.clearSelection(SELECTION_CATEGORIES.BRAND)}/>}
                                {selectedColor !== DEFAULT_SELECT_VALUE &&
                                <Chip label={selectedColor} className={classes.chip}
                                      onDelete={() => this.props.clearSelection(SELECTION_CATEGORIES.COLOR)}/>}
                            </div>
                            <Button variant="contained" color="primary" className={classes.button}
                                    onClick={this.props.resetVehicleForm}>
                                Reset Form
                            </Button>
                        </Grid>
                        <Grid item xs={12} md={6}>
                            {vehicleImage !== null &&
                                <img className={classes.vehicleImage} src={vehicleImage} alt={selectedBrand} />
                            }
                        </Grid>
                    </Grid>
                </Fragment>
                }
            </div>
        );
    }
}

ResultsDisplay.propTypes = {
    classes: PropTypes.object.isRequired,
    selectedType: PropTypes.string,
    selectedBrand: PropTypes.string,
    selectedColor: PropTypes.string,
    vehicleImage: PropTypes.string,
    formHasSelections: PropTypes.bool,
    resetVehicleForm: PropTypes.func,
    clearSelection: PropTypes.func,
};

const mapStateToProps = state => {
    return {
        formHasSelections: getMeisterFormHasSelections(state),
        selectedType: getCurrentVehicleType(state),
        selectedBrand: getCurrentVehicleBrand(state),
        selectedColor: getCurrentVehicleColor(state),
        vehicleImage: getCurrentVehicleImage(state)
    };
};

const mapDispatchToProps = {
    resetVehicleForm,
    clearSelection,
};

export default withStyles(styles)(connect(
    mapStateToProps,
    mapDispatchToProps
)(ResultsDisplay));
