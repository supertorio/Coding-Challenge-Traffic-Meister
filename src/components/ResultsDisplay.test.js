import React from 'react';
import {ResultsDisplay} from './ResultsDisplay';

describe('<ResultsDisplay />', () => {

    const classesMock = {};
    let selectedType, selectedBrand, selectedColor, formHasSelections;

    beforeEach(()=> {
        selectedType = 'car';
        selectedBrand = 'Bugatti Veyron';
        selectedColor = 'blue';
        formHasSelections = true;
    });

    it('should match the snapshot', () => {
        const wrapper = shallow(
            <ResultsDisplay classes={classesMock}
                            selectedType={selectedType}
                            selectedBrand={selectedBrand}
                            selectedColor={selectedColor}
                            formHasSelections={true} />,
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should display selected values', () => {
        const wrapper = mount(
            <ResultsDisplay classes={classesMock}
                            selectedType={selectedType}
                            selectedBrand={selectedBrand}
                            selectedColor={selectedColor}
                            formHasSelections={true} />,
        );

        expect(wrapper.find('Chip').at(0).prop('label')).toMatch(selectedType);
        expect(wrapper.find('Chip').at(1).prop('label')).toMatch(selectedBrand);
        expect(wrapper.find('Chip').at(2).prop('label')).toMatch(selectedColor);
    });

    it('should trigger form reset on button click', () => {
        const resetFormMock = jest.fn();

        const wrapper = mount(
            <ResultsDisplay classes={classesMock}
                            selectedType={selectedType}
                            selectedBrand={selectedBrand}
                            selectedColor={selectedColor}
                            formHasSelections={true}
                            resetVehicleForm={resetFormMock} />,
        );

        wrapper.find('Button').at(0).simulate('click');
        expect(resetFormMock.mock.calls.length).toBe(1);
    });

    it('should trigger clear selection on chip click', () => {
        const clearSelectionMock = jest.fn();

        const wrapper = mount(
            <ResultsDisplay classes={classesMock}
                            selectedType={selectedType}
                            selectedBrand={selectedBrand}
                            selectedColor={selectedColor}
                            formHasSelections={true}
                            clearSelection={clearSelectionMock} />,
        );

        wrapper.find('Chip').at(0).find('svg').at(0).simulate('click');
        expect(clearSelectionMock.mock.calls.length).toBe(1);
    });

});
