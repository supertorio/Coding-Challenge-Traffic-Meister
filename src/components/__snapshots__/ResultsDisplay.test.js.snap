// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResultsDisplay /> should match the snapshot 1`] = `ShallowWrapper {}`;
